packages/nimble-components/src/menu-button/index.ts

```diff
@@ -162,16 +162,52 @@ export class MenuButton extends FoundationElement implements ButtonPattern {
     public menuKeyDownHandler(e: KeyboardEvent): boolean {
         switch (e.key) {
             case keyEscape:
-                this.open = false;
+                this.setOpen(false);
                 this.toggleButton!.focus();
                 return false;
             default:
                 return true;
         }
     }
 
+    private setOpen(newValue: boolean): void {
+        if (this.open === newValue) {
+            return;
+        }
+
+        const eventDetail: MenuButtonBeforeToggleEventDetail = {
+            oldState: this.open,
+            newState: newValue
+        };
+        this.$emit('beforetoggle', eventDetail);
+
+        this.open = newValue;
+    }
+
     private get menu(): HTMLElement | undefined {
-        return this.slottedMenus?.length ? this.slottedMenus[0] : undefined;
+        // Get the menu that is slotted within the menu-button, taking into account
+        // that it may be nested within multiple 'slot' elements, such as when used
+        // within a table.
+        if (!this.slottedMenus?.length) {
+            return undefined;
+        }
+
+        let currentItem = this.slottedMenus[0];
+        while (currentItem) {
+            if (currentItem.getAttribute('role') === 'menu') {
+                return currentItem;
+            }
+
+            if (currentItem?.nodeName === 'SLOT') {
+                currentItem = (
+                    currentItem as HTMLSlotElement
```

Comment:
Can make a simple private type guard with a type predicate to clean-up this check: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates